### PR TITLE
Improve test_reference_numerics

### LINKED ISSUE #50749: nan comparision for complex numbers
Currently, the complex number comparison is done by comparing real and imaginary separately. This is not always correct, because, for example, for `x = -inf+nanj`, on CUDA 11.2, `asin(x) = nan-infj` and on gcc 10.2, `asin(x) = nan+infj`. They are both correct and I don't think one is more correct than the other. However, this way of comparing causes two test failures on my machine:
```
FAILED test/test_unary_ufuncs.py::TestUnaryUfuncsCUDA::test_reference_numerics_asin_cuda_complex128 - AssertionError: False is not true : Scalars failed to compare as equal! Comparing the imaginary part -inf and inf gives a difference of inf, but the allowed difference...
FAILED test/test_unary_ufuncs.py::TestUnaryUfuncsCUDA::test_reference_numerics_asin_cuda_complex64 - AssertionError: False is not true : Scalars failed to compare as equal! Comparing the imaginary part -inf and inf gives a difference of inf, but the allowed difference ...
```

In https://github.com/pytorch/pytorch/pull/50550, I relaxed the nan comparison at `test_unary_ufuncs.py` to
> as long as any of the real or imag is nan, then the whole number is considered as nan and considered as equal. That is `nan + 2j == 3 + nan j`.

However, as @ngimel pointed out in https://github.com/pytorch/pytorch/pull/50550#issuecomment-761905289, this is too relaxed  to make the whole `test_unary_ufuncs.py` compare `nan`s in the relaxed mode. So we need a more detailed plan about relaxed comparison and where should it be used. And this issue is for the purpose of this discussion.

cc: @mruberry 

cc @ezyang @anjali411 @dylanbespalko @mruberry

### PR DESCRIPTION
**Pull request by zasdfgbnm**
Fixes https://github.com/pytorch/pytorch/issues/50749
ci-all version of https://github.com/pytorch/pytorch/pull/50550

### CODE REVIEW

**Comment by zasdfgbnm** [comment]
I think this should be ready? There are still some old skip on "test_reference_numerics", which is a noop now, but I want to keep it in this PR and will open a new PR removing it, doing so would make it easier if we see some further failures after landing.

File: test/test_unary_ufuncs.py [hunk 1]
```diff
@@ -58,8 +58,42 @@
                      -4988429.2, 4988429.2,
                      -1e20, 1e20)
 _float_extremals = (float('inf'), float('-inf'), float('nan'))
+medium_length = 812
+large_size = (1029, 917)
 
 
+def generate_tensors_from_vals(vals, device, dtype, domain):
```
Comment (mruberry) [inline test/test_unary_ufuncs.py:65]: Add a comment describing this function, its inputs and outputs

File: test/test_unary_ufuncs.py [hunk 2]
```diff
@@ -58,8 +58,42 @@
                      -4988429.2, 4988429.2,
                      -1e20, 1e20)
 _float_extremals = (float('inf'), float('-inf'), float('nan'))
+medium_length = 812
```
Comment (mruberry) [inline test/test_unary_ufuncs.py:61]: _medium_length and _large_size

File: test/test_unary_ufuncs.py [hunk 3]
```diff
@@ -58,8 +58,42 @@
...
+large_size = (1029, 917)
 
 
+def generate_tensors_from_vals(vals, device, dtype, domain):
+    offset = 63
+
+    assert large_size[1] > (medium_length + offset)
```
Comment (mruberry) [inline test/test_unary_ufuncs.py:68]: Comment here for what these assertions are validating

File: test/test_unary_ufuncs.py [hunk 4]
```diff
@@ -58,8 +58,42 @@
...
+    large_tensor = make_tensor(large_size, device=device, dtype=dtype, low=domain[0], high=domain[1])
+
+    # Inserts the vals at an odd place
+    large_tensor[57][offset:offset + len(vals)] = torch.tensor(vals, device=device, dtype=dtype)
+
+    # Takes a medium sized copy of the large tensor containing vals
+    medium_tensor = large_tensor[57][offset:offset + medium_length]
+
+    # Constructs small tensors (4 elements)
+    small_tensors = (t for t in torch.split(medium_tensor, 4))
```
Comment (mruberry) [inline test/test_unary_ufuncs.py:82]: I think we can get rid of small_tensors now. The large tensor should CPU test vectorization (including an unvectorized tail due to its odd size), and the scalar tensors should test the unvectorized CPU code path. This change is not directly related to this PR but because we're now creating large_tensor twice it'll be nice to reclaim some perf.

File: test/test_unary_ufuncs.py [hunk 5]
```diff
@@ -112,31 +134,42 @@ def generate_numeric_tensors(device, dtype, *,
...
+    return generate_tensors_from_vals(vals, device, dtype, domain)
 
-    # Constructs the large tensor containing vals
-    large_tensor = make_tensor(large_size, device=device, dtype=dtype, low=domain[0], high=domain[1])
 
-    # Inserts the vals at an odd place
-    large_tensor[57][offset:offset + len(vals)] = torch.tensor(vals, device=device, dtype=dtype)
+def generate_numeric_tensors_hard(device, dtype, *,
+                                  domain=(None, None)):
+    is_signed_integral = dtype in (torch.int8, torch.int16, torch.int32, torch.int64)
```
Comment (mruberry) [inline test/test_unary_ufuncs.py:144]: Alternatively it could return an empty tuple for other dtypes.

File: test/test_unary_ufuncs.py [hunk 6]
```diff
@@ -112,31 +134,42 @@ def generate_numeric_tensors(device, dtype, *,
...
 
-    # Constructs scalar tensors
-    scalar_tensors = (t.squeeze() for t in torch.split(medium_tensor, 1))
 
-    # Tensors with no elements
-    empty_sizes = ((0,), (0, 3, 3), (1, 0, 5), (6, 0, 0, 0), (3, 0, 1, 0))
-    empty_tensors = (torch.empty(size, device=device, dtype=dtype) for size in empty_sizes)
+def generate_numeric_tensors_extremal(device, dtype, *,
+                                      domain=(None, None)):
+    assert dtype.is_floating_point or dtype.is_complex
```
Comment (mruberry) [inline test/test_unary_ufuncs.py:161]: Same comment here about an option to return an empty tuple instead. I'll need to keep reading to see which is preferable.

File: test/test_unary_ufuncs.py [hunk 7]
```diff
@@ -112,31 +134,42 @@ def generate_numeric_tensors(device, dtype, *,
...
+
+    vals = []
+    if dtype.is_floating_point:
+        vals = _float_extremals
+    elif dtype.is_complex:
+        vals = tuple(complex(x, y) for x, y in chain(product(_float_extremals, _float_extremals),
+                                                     product(_float_vals, _float_extremals),
+                                                     product(_float_extremals, _float_vals)))
+
+    return generate_tensors_from_vals(vals, device, dtype, domain)
```
Comment (mruberry) [inline test/test_unary_ufuncs.py:171]: This is OK for now, but I would prefer this test generate the extremal vals and just make one scalar tensor for each of them, plus take a large tensor that has NaNs, -infs, and infs scattered throughout it instead of in one concentrated section. A simple helper to "scatter" these values could be used. A density of 1/20 elements could also be used for each of them in the large tensor.

File: test/test_unary_ufuncs.py [hunk 8]
```diff
@@ -296,7 +318,43 @@ def test_reference_numerics(self, device, dtype, op):
...
+    @ops(unary_ufuncs, allowed_dtypes=floating_and_complex_types_and(torch.bfloat16, torch.half))
+    def test_reference_numerics_extremal(self, device, dtype, op):
+        handles_extremals = (op.handles_complex_extremals if
+                             dtype in (torch.cfloat, torch.cdouble) else op.handles_extremals)
+        if not handles_extremals:
+            raise self.skipTest("This op does not handle extremal values")
+
+        tensors = generate_numeric_tensors_extremal(device, dtype,
+                                                    domain=op.domain)
+        equal_nan = "relaxed" if device.startswith('cuda') else True
```
Comment (mruberry) [inline test/test_unary_ufuncs.py:356]: Add a comment here with a link to the relevant issue

File: torch/testing/__init__.py [hunk 9]
```diff
@@ -106,8 +114,19 @@ def _compare_tensors_internal(a: torch.Tensor, b: torch.Tensor, *, rtol, atol, e
     # Compares complex tensors' real and imaginary parts separately.
     # (see NOTE Test Framework Tensor "Equality")
     if a.is_complex():
+        a = a.clone()
```
Comment (mruberry) [inline torch/testing/__init__.py:117]: This clone's only needed if equal_nan is "relaxed", right? 

Also this `a_real = a.real` nonsense is just an artifact of a more complicated workaround that's no longer necessary. That was fixed in this PR https://github.com/pytorch/pytorch/pull/39065/files, so would you please get rid of a_real and a_imag and b_real and b_imag? We should just a.real and a.imag and etc.

Comment (mruberry) [reply torch/testing/__init__.py:117]: With those changes I think this simplifies to:

```
if equal_nan == "relaxed":
  a = a.clone()
  b = b.clone()
  a.real[a.imag.isnan()] = math.nan
  a.imag[a.real.isnan()] = math.nan
  b.real[b.imag.isnan()] = math.nan
  b.imag[b.real.isnan()] = math.nan
```

File: torch/testing/__init__.py [hunk 10]
```diff
@@ -129,7 +146,7 @@ def _compare_tensors_internal(a: torch.Tensor, b: torch.Tensor, *, rtol, atol, e
         return (True, None)
 
     # All other comparisons use torch.allclose directly
-    if torch.allclose(a, b, rtol=rtol, atol=atol, equal_nan=equal_nan):
+    if torch.allclose(a, b, rtol=rtol, atol=atol, equal_nan=bool(equal_nan)):
```
Comment (mruberry) [inline torch/testing/__init__.py:149]:
```
equal_nan = (equal_nan in {"relaxed", True})
```

for maintainability/clarity

File: torch/testing/__init__.py [hunk 11]
```diff
@@ -140,7 +157,7 @@ def _compare_tensors_internal(a: torch.Tensor, b: torch.Tensor, *, rtol, atol, e
 
     # Masks close values
     # NOTE: this avoids (inf - inf) oddities when computing the difference
-    close = torch.isclose(a_flat, b_flat, rtol, atol, equal_nan)
+    close = torch.isclose(a_flat, b_flat, rtol, atol, bool(equal_nan))
```
Comment (mruberry) [inline torch/testing/__init__.py:160]: Same change to `bool(equal_nan)` here (or do them both at once in a local variable)

File: torch/testing/__init__.py [hunk 12]
```diff
@@ -163,14 +180,14 @@ def _compare_tensors_internal(a: torch.Tensor, b: torch.Tensor, *, rtol, atol, e
 
 # Checks if two scalars are equal(-ish), returning (True, None)
 # when they are and (False, debug_msg) when they are not.
-def _compare_scalars_internal(a, b, *, rtol: float, atol: float, equal_nan: bool) -> _compare_return_type:
+def _compare_scalars_internal(a, b, *, rtol: float, atol: float, equal_nan: Union[str, bool]) -> _compare_return_type:
     def _helper(a, b, s) -> _compare_return_type:
         # Short-circuits on identity
-        if a == b or (equal_nan and a != a and b != b):
+        if a == b or (bool(equal_nan) and a != a and b != b):
```
Comment (mruberry) [inline torch/testing/__init__.py:186]: This also needs to assert that equal_nan is a valid value and test if that value is "Relaxed" or True directly (not rely on strings being converted to True)

File: torch/testing/__init__.py [hunk 13]
```diff
@@ -199,6 +216,12 @@ def _helper(a, b, s) -> _compare_return_type:
         a = complex(a)
         b = complex(b)
 
+        if equal_nan == "relaxed":
```
Comment (mruberry) [inline torch/testing/__init__.py:219]: This seems like it will create a confusing error message in cases like 5 + 0j being compared with 5 + NaNj, because the latter number will become NaN + NaNj and the real part will compare as different, not the imaginary part. 

Instead, when equal_nan is relaxed just short-circuit future checks like this:

```
if a != a and b != b:
  return (True, None)
```

**Review by mruberry — COMMENTED** [review]
This is a very good and thorough PR implementing a significant improvement to the testing of unary functions.

I've made some inline comments. They don't change the overall structure but there are a few interesting issues to resolve. When they're fixed and this is rebased I'll review ASAP and hold additional OpInfo PRs until this is landed to minimize the chance this needs to be rebased again. 

Looking forward to the update!

File: test/test_unary_ufuncs.py [hunk 14]
```diff
@@ -58,8 +58,42 @@
...
 
+# Returns generator of tensors of different sizes filled with values in domain
+# and with intested region filled with `vals`. This will help test different code
+# paths for the given vals
+def generate_tensors_from_vals(vals, device, dtype, domain):
+    offset = 63
+
+    assert _large_size[1] > (_medium_length + offset)  # large tensor should be large enough
+    assert len(vals) < _medium_length  # medium tensor should contain all vals
+    assert _medium_length % 4 == 0  # ensure vectorized code coverage
```
Comment (zasdfgbnm) [inline test/test_unary_ufuncs.py:73]: Is my comment correct here?

**Comment by zasdfgbnm** [comment]
@mruberry I have resolved your reviews

File: test/test_unary_ufuncs.py (hunk 14 above)
Comment (mruberry) [reply test/test_unary_ufuncs.py:73]: Kind of? Any contiguous innermost dimensions will be vectorized. This condition ensures there will be no non-vectorized tail.

**Review by mruberry — APPROVED** [review]
Thanks @zasdfgbnm! This should let us get much better complex correctness coverage!
